server/routes/linkExistingAddressToChain.ts

```diff
@@ -0,0 +1,141 @@
+import { Request, Response, NextFunction } from 'express';
+import Sequelize from 'sequelize';
+import crypto from 'crypto';
+import { ADDRESS_TOKEN_EXPIRES_IN } from '../config';
+
+const { Op } = Sequelize;
+
+export const Errors = {
+  NeedAddress: 'Must provide address',
+  NeedChain: 'Must provide chain',
+  NeedOriginChain: 'Must provide original chain',
+  NeedLoggedIn: 'Must be logged in',
+  NotVerifiedAddressOrUser: 'Not verified address or user',
+  InvalidChain: 'Invalid chain',
+};
+
+const linkExistingAddressToChain = async (models, req: Request, res: Response, next: NextFunction) => {
+  if (!req.body.address) {
+    return next(new Error(Errors.NeedAddress));
+  }
+  if (!req.body.chain) {
+    return next(new Error(Errors.NeedChain));
+  }
+  if (!req.body.originChain) {
+    return next(new Error(Errors.NeedOriginChain));
+  }
+  if (!req.user?.id) {
+    return next(new Error(Errors.NeedLoggedIn));
+  }
+
+  const userId = req.user.id;
+
+  const chain = await models.Chain.findOne({
+    where: { id: req.body.chain }
+  });
+  if (!chain) {
+    return next(new Error(Errors.InvalidChain));
+  }
+
+  // check if the original address is verified and is owned by the user
+  const originalAddress = await models.Address.scope('withPrivateData').findOne({
+    where: { chain: req.body.originChain, address: req.body.address, user_id: userId, verified: { [Op.ne]: null } }
+  });
+
+  if (!originalAddress) {
+    return next(new Error(Errors.NotVerifiedAddressOrUser));
+  }
+
+  // check if the original address's token is expired. refer edge case 1)
+  let verificationToken = originalAddress.verification_token;
+  let verificationTokenExpires = originalAddress.verification_token_expires;
+  const isOriginalExpired = verificationTokenExpires && +verificationTokenExpires <= +(new Date());
+
+  if (!isOriginalExpired) {
+    const chains = await models.Chain.findAll({
+      where: { base: chain.base }
+    });
+
+    verificationToken = crypto.randomBytes(18).toString('hex');
+    verificationTokenExpires = new Date(+(new Date()) + ADDRESS_TOKEN_EXPIRES_IN * 60 * 1000);
+
+    await models.Address.update({
+      verification_token: verificationToken,
+      verification_token_expires: verificationTokenExpires
+    }, {
+      where: {
+        user_id: originalAddress.user_id,
+        address: req.body.address,
+        chain: { [Op.in]: chains.map((ch) => ch.id) }
+      }
+    });
+  }
```

Comment:
Can you describe the intended behavior here?

It looks like you're checking if the verification token is valid, and then if it is valid, resetting it into the future again?

Reply:
I named the variable wrongly. I've changed `isOriginalExpired` to `isOriginalTokenValid`.

If the verification token is not valid, it resets it into the future.

Reply:
```
const isOriginalTokenValid = verificationTokenExpires && +verificationTokenExpires <= +(new Date());

  if (!isOriginalTokenValid) {
```